File: backend/repository/OrderRepository.js
const Database = require('../database/Database');
const { DatabaseError, NotFoundError, ValidationError } = require('../utils/Errors');
const RepositroyAbstract = require('./RepositoryFactory');

class OrderRepository extends RepositroyAbstract{
    static tableName = 'orders';
    static async insert(order,  {transaction} = {}){
        console.log(transaction)
        const pool = transaction || Database.getPool();
        const query = `
            INSERT INTO orders (
                user_id, 
                stock_id, 
                order_type,      
                order_side,      
                price, 
                quantity,
                remaining_quantity,
                status
            ) VALUES ($1, $2, $3, $4, $5, $6, $6, $7) RETURNING order_id`;
        
        const values = [
            order.user_id, 
            order.stock_id, 
            order.order_type,    // Limit/Market
            order.order_side,    // Buy/Sell
            order.price, 
            order.quantity,
            'pending'            // 初始狀態
        ];
        try{
            const result = await pool.query(query, values);
            return result.rows[0].order_id;
        }catch(error){
            console.error('Error adding order:', error);
            throw new DatabaseError(`訂單資料庫新增錯誤: ${error.message}`);
        }
    }

    static async delete(order_id,  {transaction}  = {}){
        const pool = transaction || Database.getPool();
        const query = 'DELETE FROM orders WHERE order_id = $1';
        const values = [order_id];
        try{
            await pool.query(query, values);
        }catch(error){
            console.error('Error deleting order:', error);
            throw new DatabaseError(`訂單資料庫刪除錯誤: ${error.message}`);
        }
    }

    static async update(data, { transaction } = {}) {
        const pool = transaction || Database.getPool();
        const query = `
            UPDATE orders 
            SET price = COALESCE($1, price),
                remaining_quantity = COALESCE($2, remaining_quantity),
                status = COALESCE($3, status)
            WHERE order_id = $4
            RETURNING *`;
            
        const values = [
            data.price, 
            data.remaining_quantity,
            data.status,
            data.order_id
        ];
        try{
            const result = await pool.query(query, values);
            return result.rows[0];
        }catch(error){
            console.error('Error updating order:', error);
            throw new DatabaseError(`訂單資料庫更新錯誤: ${error.message}`);
        }
    }
}

module.exports = OrderRepository;

File: backend/repository/orderRepository.py
from backend.database.database import Database
from backend.utils.errors import DatabaseError
from backend.repository.repositoryAbs import RepositoryAbstract
from backend.models.order.order import OrderType
from typing import Dict

class OrderRepository(RepositoryAbstract):
    table_name = 'orders'
    key = 'order_id'
    @classmethod
    async def insert(cls, order, transaction=None):
        pool = transaction or Database.get_pool()
        query = """
            INSERT INTO orders (
                user_id, 
                stock_id, 
                order_type,      
                order_side,      
                quantity,
                remaining_quantity,
                status,
                price
            ) VALUES ($1, $2, $3, $4, $5, $6, $7, $8) RETURNING *"""
        
        values = [
            order.user_id, 
            order.stock_id, 
            order.order_type,    # Limit/Market
            order.order_side,    # Buy/Sell
            order.quantity,
            order.remaining_quantity,
            order.status,
            order.price if order.order_type == OrderType.Limit else None  # 只有限價單才設置價格
        ]

        try:
            result = await pool.fetchrow(query, *values)
            return result['order_id']
        except Exception as error:
            print('Error adding order:', error)
            raise DatabaseError(f"訂單資料庫新增錯誤: {str(error)}")

    @classmethod
    async def delete(cls, order_id, transaction=None):
        pool = transaction or Database.get_pool()
        query = 'DELETE FROM orders WHERE order_id = $1'
        values = [order_id]
        try:
            await pool.execute(query, *values)
        except Exception as error:
            print('Error deleting order:', error)
            raise DatabaseError(f"訂單資料庫刪除錯誤: {str(error)}")
